compiler/codegen/OMRCodeGenerator.cpp

```diff
@@ -358,6 +363,164 @@ OMR::CodeGenerator::generateCodeFromIL()
+      //
+      lastWarmBlock = block;
+      }
+
+   lastWarmBlock->setIsLastWarmBlock();
+
+   if (comp->getOption(TR_TraceCG))
+      {
+      traceMsg(comp, "%s Last warm block is block_%d\n", SPLIT_WARM_COLD_STRING, lastWarmBlock->getNumber());
+
+      if (numColdBlocks > 0)
+         traceMsg(comp, "%s Moved to cold code cache %d out of %d cold blocks (%d%%)\n",
+                           SPLIT_WARM_COLD_STRING,
+                           numColdBlocks - numNonOutlinedColdBlocks,
+                           numColdBlocks,
+                           (numColdBlocks - numNonOutlinedColdBlocks)*100/numColdBlocks);
+      }
+
+   // If the last tree in the last warm block is not a TR_goto, insert a goto tree
+   // at the end of the block.
+   // If there is a following block the goto will branch to it so that when the
+   // code is split any fall-through will go to the right place.
+   // If there is no following block the goto will branch to the first block; in
+   // this case the goto should never be reached, it is there only to
+   // make sure that the instruction following the last real treetop will be in
+   // warm code, so if it is a helper call (e.g. for a throw) the return address
+   // is in this method's code.
+   //
+   if (lastWarmBlock->getNumberOfRealTreeTops() == 0)
+      tt = lastWarmBlock->getEntry();
+   else
+      tt = lastWarmBlock->getLastRealTreeTop();
+
+   node = tt->getNode();
+
+   if (!(node->getOpCode().isGoto() ||
+         node->getOpCode().isJumpWithMultipleTargets() ||
+         node->getOpCode().isReturn()))
+      {
+      // Find the block to be branched to
+      //
+      TR::TreeTop * targetTreeTop = lastWarmBlock->getExit()->getNextTreeTop();
+
+      if (targetTreeTop)
+         // Branch to following block. Make sure it is not marked as an
+         // extension block so that it will get a label generated.
+         //
+         targetTreeTop->getNode()->getBlock()->setIsExtensionOfPreviousBlock(false);
+      else
+         // Branch to the first block. This will not be marked as an extension
+         // block.
+         //
+         targetTreeTop = comp->getStartBlock()->getEntry();
+
+      // Generate the goto and insert it into the end of the last warm block.
+      //
+      TR::TreeTop *gotoTreeTop = TR::TreeTop::create(comp, TR::Node::create(node, TR::Goto, 0, targetTreeTop));
+
+      // Move reg deps from BBEnd to goto
+      //
+      TR::Node *bbEnd = lastWarmBlock->getExit()->getNode();
+
+      if (bbEnd->getNumChildren() > 0)
+         {
+         TR::Node *glRegDeps = bbEnd->getChild(0);
+
+         gotoTreeTop->getNode()->setNumChildren(1);
+         gotoTreeTop->getNode()->setChild(0, glRegDeps);
+
+         bbEnd->setChild(0,NULL);
+         bbEnd->setNumChildren(0);
+         }
+
+      tt->insertAfter(gotoTreeTop);
+
+      // Update CFG
+      //
+      TR::CFG *cfg = comp->getFlowGraph();
+      cfg->addEdge(lastWarmBlock, comp->getStartBlock());
```

Comment:
Shouldn't this add an edge to `targetTreeTop`, since the goto isn't always going to the start block?

Reply:
I think I made a mistake with the last fix. It's better not to insert any edge. It's only needed for an unreachable `goto `so conceptually does not affect CFG. I think I will remove the code above.